Source/Atomic/UI/UIFinderWindow.cpp

```diff
@@ -0,0 +1,618 @@
+{
+    if (!widget_)
+        return;
+
+    if ( ((TBFinderWindow*)widget_)->Show(title.CString(), preset.CString(), dimmer, width, height) )
+    {
+        CreateBookmarks();
+        PresetCurrentPath(preset);
+        UpdateUiPath();
+        UpdateUiList();
+        UpdateUiResult();
+    }
+}
+
+void UIFinderWindow::FindPath(const String& title, const String& preset, int dimmer, int width, int height)
+{
+    if (!widget_)
+        return;
+
+    finderMode = 1;
+    
+    if ( ((TBFinderWindow*)widget_)->Show(title.CString(), preset.CString(), dimmer, width, height) )
+    {
+        UIWidget *reswid = GetResultWidget();
+        if (reswid) reswid->SetVisibility( UI_WIDGET_VISIBILITY_INVISIBLE );
+        CreateBookmarks();
+        PresetCurrentPath(preset);
+        UpdateUiPath();
+        UpdateUiList();
+        UpdateUiResult();
+    }
+}
+
+bool UIFinderWindow::OnEvent(const tb::TBWidgetEvent &ev)
+{
+
+    if ( ev.type == EVENT_TYPE_CHANGED && ev.target && ((uint32)ev.target->GetID()) == 1 ) 
+    {
+        FileSystem* filesystem = GetSubsystem<FileSystem>();
+        UIWidget *pathwidget = GetPathWidget();  // paste or type in currentpath widget
+        if(pathwidget)
+        {
+            if ( filesystem->DirExists (pathwidget->GetText())) 
+            {
+                if ( pathwidget->GetText() != currentPath ) 
+                {
+                    resultPath = "";
+                    SetCurrentPath (pathwidget->GetText());
+                    UpdateUiList();
+                    UpdateUiResult();
+                }
+            }
+        }
+        return true;
+    }
+
+   if ( ev.type == EVENT_TYPE_POINTER_UP && ev.target && ((uint32)ev.target->GetID()) == 2 ) // go up
+    {
+        GoFolderUp();
+        return true;
+    }
+
+    if ( ev.type == EVENT_TYPE_POINTER_UP && ev.target && ((uint32)ev.target->GetID()) == 3 ) // add bookmark request
+    {
+        // this check is necessary because you can kill the bookmark window with the "X" and
+        // this kills the UI but not the newBookmarkPtr to it, and it doesnt get cleaned up properly.
+        if ( newBookmarkPtr.NotNull() )
+        {
+            newBookmarkPtr->UnsubscribeFromAllEvents();
+            newBookmarkPtr.Reset();
+        }
+        newBookmarkPtr = new UIPromptWindow(context_, this, "createbookmark",  true);
+        SubscribeToEvent(newBookmarkPtr, E_UIPROMPTCOMPLETE, ATOMIC_HANDLER(UIFinderWindow, HandleCreateBookmark ));
+
+        String prospect = "";
+        char delim = '/';
+        Vector <String> tokens = currentPath.Split(delim, false);
+        prospect = tokens[ tokens.Size()-1 ]; // get the last folder name as preset
+    
+        newBookmarkPtr->Show("Create New Bookmark", "Enter a name for the new bookmark", prospect );
+        return true;
+    }
+
+    if ( ev.type == EVENT_TYPE_POINTER_UP && ev.target && ((uint32)ev.target->GetID()) == 4 ) // add folder request
```

Comment:
I was using TB as an example, admittedly a poor example. I will make some defines, sadly I cant embed the defines in the tb ui definition string.
And dont get me started on the wondrous TB id system, too late.  A great system, use a name (hashed to a number) or a number, does not have to be unique, and all global. Great for toy programs, sux when you scale up.